qiita_pet/templates/study_ajax/processing_study.html

```diff
@@ -0,0 +1,164 @@
+<script type="text/javascript">
+  //the network object itself, + nodes list and edge list
+  var proc_network = null;
+  var nodes = new vis.DataSet([{id: 1, label: "{{artifact_type}}", group: 'type'}]);
+  var edges = new vis.DataSet();
+  //current commands available for selected nodes
+  var commands = null;
+  //currently selected nodes
+  var curr_nodes = [1];
+
+  function add_command(type_nodes) {
+    var command_info = commands[$("#command").val()];
+    //TODO: add job callback
+    var command_node = nodes.length + 1;
+    // Add edges to the new command node, then the new command node
+    for(var i=0;i<type_nodes.length;i++) {
+      edges.add({ id: edges.length + 1, from: type_nodes[i], to: command_node});
+    }
+    nodes.add({ id: command_node, label: command_info.command, group: 'command', command_id: command_info.id});
+    // Add edges for types created by command
+    for(i=0;i<command_info.output.length;i++) {
+      var node_id = nodes.length + 1;
+      edges.add({ id: edges.length + 1, from: command_node, to: node_id});
+      nodes.add({ id: node_id, label: command_info.output[i], group: 'type', command_id: null});
+    }
+    proc_network.redraw();
+  }
+
+  function remove_command(node_id) {
+    //TODO: delete job callback
+    if(node_id.length > 1) { alert("Can only delete one command at a time."); return;}
```

Comment:
This is done more than once. Perhaps a function (function validate_node_id) that also returns an int?